### PR TITLE
Simplify annotations in `terminator` module

### PR DESCRIPTION
**Pull request by eukaryo**
## Motivation

Related to Issue https://github.com/optuna/optuna/issues/4508.

I will work on making changes to the `terminator`, and want to resolve the issue in the module as a preliminary action.

## Description of the changes

I removed `List`, `Dict`, and `Optional` to typing from `terminator` module, and refactored the code.

### CODE REVIEW

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Comment by HideakiImamura** [comment]
@nabenabe0928 Could you review this PR?

File: optuna/terminator/terminator.py [hunk 1]
```diff
@@ -101,8 +100,8 @@ class Terminator(BaseTerminator):
 
     def __init__(
         self,
-        improvement_evaluator: Optional[BaseImprovementEvaluator] = None,
-        error_evaluator: Optional[BaseErrorEvaluator] = None,
+        improvement_evaluator: BaseImprovementEvaluator | None = None,
+        error_evaluator: BaseErrorEvaluator | None = None,
         min_n_trials: int = DEFAULT_MIN_N_TRIALS,
     ) -> None:
```
Comment (nabenabe0928) [inline optuna/terminator/terminator.py:106]:
```suggestion
    ):
```

File: optuna/terminator/improvement/evaluator.py [hunk 2]
```diff
@@ -268,7 +266,7 @@ def evaluate(
     @classmethod
     def _validate_input(
         cls,
-        trials: List[FrozenTrial],
+        trials: list[FrozenTrial],
     ) -> None:
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:270]:
```suggestion
    def _validate_input(cls, trials: list[FrozenTrial]) -> None:
```

File: optuna/terminator/improvement/evaluator.py [hunk 3]
```diff
@@ -244,7 +242,7 @@ def __init__(
 
     def evaluate(
         self,
-        trials: List[FrozenTrial],
+        trials: list[FrozenTrial],
         study_direction: StudyDirection,
     ) -> float:
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:247]:
```suggestion
    def evaluate(self, trials: list[FrozenTrial], study_direction: StudyDirection) -> float:
```

**Review by nabenabe0928 — COMMENTED** [review]
I added some comments and could you remove `-> None` in the `__init__` methods of all objects in the files you made the changes?

File: optuna/terminator/improvement/evaluator.py [hunk 4]
```diff
@@ -165,7 +163,7 @@ def _compute_standardized_regret_bound(
...
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:168]:
```suggestion
    def evaluate(self, trials: list[FrozenTrial], study_direction: StudyDirection) -> float:
```

File: optuna/terminator/improvement/evaluator.py [hunk 5]
```diff
@@ -55,7 +53,7 @@ class BaseImprovementEvaluator(metaclass=abc.ABCMeta):
     @abc.abstractmethod
     def evaluate(
         self,
-        trials: List[FrozenTrial],
+        trials: list[FrozenTrial],
         study_direction: StudyDirection,
     ) -> float:
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:58]:
```suggestion
    def evaluate(self, trials: list[FrozenTrial], study_direction: StudyDirection) -> float:
```

File: optuna/terminator/callback.py [hunk 6]
```diff
@@ -66,7 +64,7 @@ def objective(trial):
 
     def __init__(
         self,
-        terminator: Optional[BaseTerminator] = None,
+        terminator: BaseTerminator | None = None,
     ) -> None:
```
Comment (nabenabe0928) [inline optuna/terminator/callback.py:68]:
```suggestion
    def __init__(self, terminator: BaseTerminator | None = None):
```

File: optuna/terminator/terminator.py (hunk 1 above)
Comment (eukaryo) [reply optuna/terminator/terminator.py:106]: [PEP 484](https://peps.python.org/pep-0484/#the-meaning-of-annotations) says:

> the return type of `__init__` ought to be annotated with `-> None`.

File: optuna/terminator/improvement/evaluator.py (hunk 2 above)
Comment (eukaryo) [reply optuna/terminator/improvement/evaluator.py]: Agreed. I changed the code.

File: optuna/terminator/callback.py (hunk 6 above)
Comment (eukaryo) [reply optuna/terminator/callback.py]: I changed the code except that I left `-> None`.

**Review by nabenabe0928 — APPROVED** [review]
LGTM

**Pull request merged**